src/states/node.rs

```diff
@@ -971,63 +1038,131 @@ impl Node {
+                                      src: Authority<XorName>,
+                                      dst: Authority<XorName>)
+                                      -> Result<(), RoutingError> {
+        let name = match src {
+            Authority::Client { .. } => public_id.name(),
+            Authority::ManagedNode(ref name) => name,
+            _ => unreachable!(),
+        };
+        self.peer_mgr.allow_connect(name)?;
+        let their_connection_info = self.decrypt_connection_info(&encrypted_connection_info,
+                                     &box_::Nonce(nonce_bytes),
+                                     &public_id)?;
+        let peer_id = their_connection_info.id();
+        use peer_manager::ConnectionInfoReceivedResult::*;
+        match self.peer_mgr
+            .connection_info_received(src, dst, public_id, their_connection_info, message_id) {
+            Ok(Ready(our_info, their_info)) => {
+                info!("{:?} Already sent a connection info request to {:?} ({:?}); resending our \
+                      same details as a response.",
+                      self,
+                      public_id.name(),
+                      peer_id);
+                let _ = self.do_send_connection_info(our_info.to_pub_connection_info(),
+                                                     public_id,
+                                                     dst,
+                                                     src,
+                                                     Some(message_id));
+                if let Err(error) = self.crust_service.connect(our_info, their_info) {
+                    trace!("{:?} Unable to connect to {:?} - {:?}", self, src, error);
+                }
+            }
+            Ok(Prepare(token)) => {
+                self.crust_service.prepare_connection_info(token);
+            }
+            Ok(IsProxy) |
+            Ok(IsClient) |
+            Ok(IsJoiningNode) => {
+                self.send_node_identify(peer_id)?;
+                self.handle_node_identify(public_id, peer_id);
+            }
+            Ok(Waiting) | Ok(IsConnected) => (),
+            Err(error) => {
+                warn!("{:?} Failed to insert connection info from {:?} ({:?}): {:?}",
+                      self,
+                      public_id.name(),
+                      peer_id,
+                      error)
+            }
+        }
+        Ok(())
     }
 
-    fn handle_connection_info_from_node(&mut self,
-                                        conn_info: ConnectionInfo,
-                                        src_name: XorName,
-                                        dst: Authority<XorName>)
-                                        -> Result<(), RoutingError> {
-        self.peer_mgr.allow_connect(&src_name)?;
-        self.connect(conn_info, dst, Authority::ManagedNode(src_name))
+    fn handle_connection_info_response(&mut self,
+                                       encrypted_connection_info: Vec<u8>,
```

Comment:
I did consider that, and abstracted what I could to a new function to cut down on duplication, but I think there are enough differences inside the match (which makes up the bulk of the functions) to justify having them separate.